### DISCUSSION TITLE
Wysiwig builder?

### DISCUSSION THREAD

**Discussion by alexander-potemkin** [lowdefy/lowdefy#704] [Ideas and feature requests]
Hi and thank you for the product - it looks just grat!
Is there any plans to add a WYSIWIG builder? If not, any thoughts on how it could be implemented (in the best way)?

**Comment by Gervwyk** [comment]
Hi @alexander-potemkin Glad you like it 💫
When you say WYSIWIG builder, do you mean a WYSIWIG Lowdefy app builder or a WYSIWIG block to use in Lowdefy apps?

**Reply by alexander-potemkin** [reply]
Hi @Gervwyk and thank you for your prompt responce!

If the second - it’s a way to format the text in WYSIWIG way, then it’s not it =)

I liked the style and the way the system works, but writing YML file doesn’t feel too much different than to right a code in some framework for me, but I understand that YML file could be pretty much easy generated by some tool, while I drag and drop things that I need to the canvas, unless I’m missing something?

**Reply by Gervwyk** [reply]
So for now we focus on creating the best developer-first experience with Lowdefy, and that means that writing the apps should work well with dev tools and code editors so that it is easy to write apps, review changes and apply source control to apps. 

In the future, we hope to create a nice "drag and drop" build your app GUI for Lowdefy to make Lowdefy more accessible to non-developers. 

As for a WYSIWIG input block, we'd like to create a Lowdefy block as discussed here: https://github.com/lowdefy/lowdefy/discussions/492#discussioncomment-478072

In the meantime, you can create a [custom block](https://docs.lowdefy.com/custom-blocks) for this, and if you'd like to contribute to Lowdefy to make it a default block, please let me know.

**Reply by alexander-potemkin** [reply]
Ok, thank you!
And any words of wisdom, architectural advice, if I decide to implement one (drag & drop GUI for non-developers)?

**Reply by Gervwyk** [reply]
Mmm.. I think in the future, it could be built using Lowdefy, making it very easy to manage and extend. However, currently, we do not dynamically recreate blocks in the Lowdefy engine. So modifying the app config requires a page reload. Thus the easiest way to do it currently would have to be using the local file system instead of an online editor.

 I always wondered if one could code a VSCode extension to do this. With VSCode you get 70% of what would be required in the UI out of the box, so you could just focus on the GUI. It also comes with everything out of the box, tab handling, github sign in etc. So perhaps, it's the best place to start. I've not researched this, so I can't say what limitations VSCode has for extensions.

**Reply by alexander-potemkin** [reply]
I see, thanks for pointing out on that important limitation (luck of dynamic blocks recreation).

And could you please, elaborate - what do you mean, speaking about VS Code having 70% of what is needed? I understand Git integration, tabs handling, but what else is there? Shouldn’t it be easier if it resides inside the system, as opposed to the (extensible) but text editor? What am I missing?

**Reply by Gervwyk** [reply]
I think (not sure) with VSCode your extension could be a web view on the current files. So because Lowdefy allows you to work with files and decide what pieces of the config you'd like to template using the `_ref` operator, it would be nice to have full access to all basic file system features available through the UI. VSCode is also great for this.

In my opinion (and it needs more thought), the one thing required to make a nice GUI for Lowdefy inside of VSCode would just be a different view on the config structure, almost like a json explorer of the compiled config, and then highlighting the right block / element as you click on it in the app, with the required properties available to change for that specific block / element. Sort of how Webflow's GUI for editing web pages work.

But it's not a simple task, if your looking to go down this route perhaps we can have a chat about this some time to exchange some ideas? You're welcome to DM me on discord https://discord.gg/WmcJgXt
